Signal/src/ViewControllers/OWSAudioAttachmentPlayer.m

```diff
@@ -0,0 +1,173 @@
+
+@implementation OWSAudioAttachmentPlayer
+
+- (instancetype)initWithMediaAdapter:(TSVideoAttachmentAdapter *)mediaAdapter
+                  databaseConnection:(YapDatabaseConnection *)databaseConnection
+{
+    self = [super init];
+    if (!self) {
+        return self;
+    }
+
+    OWSAssert(mediaAdapter);
+    OWSAssert([mediaAdapter isAudio]);
+    OWSAssert(mediaAdapter.attachmentId);
+    OWSAssert(databaseConnection);
+
+    _mediaAdapter = mediaAdapter;
+
+    __block TSAttachment *attachment = nil;
+    [databaseConnection readWithBlock:^(YapDatabaseReadTransaction *transaction) {
+        attachment = [TSAttachment fetchObjectWithUniqueID:mediaAdapter.attachmentId transaction:transaction];
+    }];
+    OWSAssert(attachment);
+
+    if ([attachment isKindOfClass:[TSAttachmentStream class]]) {
+        self.attachmentStream = (TSAttachmentStream *)attachment;
+    }
+    OWSAssert(self.attachmentStream);
+
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(applicationDidEnterBackground:)
+                                                 name:UIApplicationDidEnterBackgroundNotification
+                                               object:nil];
+
+    return self;
+}
+
+- (void)dealloc
+{
+    [[NSNotificationCenter defaultCenter] removeObserver:self];
+
+    [self stop];
+}
+
+- (void)applicationDidEnterBackground:(NSNotification *)notification
+{
+    [self stop];
+}
+
+#pragma mark - Methods
+
+- (void)play
+{
+    OWSAssert(self.attachmentStream);
+    OWSAssert(![self.mediaAdapter isAudioPlaying]);
+
+    [ViewControllerUtils setAudioIgnoresHardwareMuteSwitch:YES];
+
+    [self.audioPlayerPoller invalidate];
+
+    self.mediaAdapter.isAudioPlaying = YES;
+    self.mediaAdapter.isPaused = NO;
+    [self.mediaAdapter setAudioIconToPause];
+
+    if (!self.audioPlayer) {
+        NSError *error;
+        self.audioPlayer = [[AVAudioPlayer alloc] initWithContentsOfURL:self.attachmentStream.mediaURL error:&error];
+        if (error) {
+            DDLogError(@"%@ error: %@", self.tag, error);
+            [self stop];
+            return;
+        }
+        self.audioPlayer.delegate = self;
+    }
+
+    [self.audioPlayer prepareToPlay];
+    [self.audioPlayer play];
+    self.audioPlayerPoller = [NSTimer scheduledTimerWithTimeInterval:.05
+                                                              target:self
+                                                            selector:@selector(audioPlayerUpdated:)
+                                                            userInfo:nil
+                                                             repeats:YES];
+}
+
+- (void)pause
+{
+    OWSAssert(self.attachmentStream);
+
+    self.mediaAdapter.isAudioPlaying = NO;
+    self.mediaAdapter.isPaused = YES;
+    [self.audioPlayer pause];
+    [self.audioPlayerPoller invalidate];
+    double current = [self.audioPlayer currentTime] / [_audioPlayer duration];
```

Comment:
nit: Objc, how do I style thee? Let me count the ways...

1. `[self.audioPayer duration]` 
2. `[_audioPlayer duration]`
3. `_audioPlayer.duration` 
4. `self.audioPlayer.duration`

I prefer property accessors wherever possible, but in any case we'd benefit from some consistency. Take your pick.